src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+            internal TelemetryListener Owner;          // The TelemetryListener this is a subscription for.  
+            internal Subscription Next;                // Linked list of subscribers
+
+            public void Dispose()
+            {
+                // TO keep this lock free and easy to analyze, the linked list is READ ONLY.   Thus we copy
+                for (;;)
+                {
+                    Subscription subscriptions = Owner.m_subscriptions;
+                    Subscription newSubscriptions = Remove(subscriptions, this);    // Make a new list, with myself removed.  
+
+                    // try to update, but if someone beat us to it, then retry.  
+                    if (Interlocked.CompareExchange(ref Owner.m_subscriptions, newSubscriptions, subscriptions) == subscriptions)
+                    {
+#if DEBUG
+                        var cur = subscriptions;
+                        while (cur != null)
+                        {
+                            if (cur.Observer == this)
+                                throw new Exception("Did not remove subscription!");
+                            cur = cur.Next;
+                        }
+#endif
+                        break;
+                    }
+                }
+            }
+
+            // Create a new linked list where 'subscription has been removed from the linked list of 'subscriptions'. 
+            private static Subscription Remove(Subscription subscriptions, Subscription subscription)
+            {
+                if (subscriptions == null)
+                {
+#if DEBUG
+                    throw new Exception("Could not find subscription");
+#else
+                    return null;
+#endif
+                }
+
+                if (subscriptions.Observer == subscription.Observer && subscriptions.IsEnabled == subscription.IsEnabled)
+                    return subscriptions.Next;
+#if DEBUG
+                // Delay a bit.  This makes it more likely that races will happen. 
+                for (int i = 0; i < 100; i++)
+                    GC.KeepAlive("");
+#endif 
+                return new Subscription() { Observer = subscriptions.Observer, Owner = subscriptions.Owner, IsEnabled = subscriptions.IsEnabled,  Next = Remove(subscriptions.Next, subscription) };
+            }
+        }
+
+        Subscription m_subscriptions;
+        TelemetryListener m_next;               // We keep a linked list of all NotificationListeners (s_allListeners)
+        bool m_disposed;                        // Has Dispose been called?
+
+        static Action<TelemetryListener> s_allListenersCallback;    // The list of clients to call back when a new TelemetryListener is created.  
+        static TelemetryListener s_allListeners;                    // As a service, we keep track of all instances of NotificationListeners.  
+        static TelemetryListener s_default = new TelemetryListener("TelemetryListener.Default");
```

Comment:
Nit: the style guidelines for corefx also have visibility always specified even if it's the default, e.g. all of these instance and static fields should be prefaced with ```private```.